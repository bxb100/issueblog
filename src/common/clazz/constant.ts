export class Constant {
    static readonly FIXED_LINKS = 'Links'
    static readonly FIXED_TOP = 'Top'
    static readonly FIXED_TODO = 'Todo'
    static readonly FIXED_RECENT = 'Recent'
    static readonly AGG_EACH_LABEL = 'Label'

    readonly header: string

    constructor(header: string) {
        this.header = header
    }

    convertBlogContent(map: Map<string, string>): string {
        let content: string = this.header
        content += map.get(Constant.FIXED_LINKS) || ''
        content += map.get(Constant.FIXED_TOP) || ''
        content += map.get(Constant.FIXED_RECENT) || ''
        content += map.get(Constant.AGG_EACH_LABEL) || ''
        content += map.get(Constant.FIXED_TODO) || ''
        return content
    }
}
