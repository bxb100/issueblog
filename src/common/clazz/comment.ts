import {GithubKit} from './github-kit'
import {IComment} from '../interface/comment'
import {IUser} from '../interface/user'
import {Reactions} from '../interface/reactions'

export class Comment implements IComment {
    readonly id: number
    readonly user: IUser | null
    readonly reactions: Reactions | undefined
    readonly body: string | undefined

    constructor(comment: Readonly<IComment>) {
        this.id = comment.id
        this.user = comment.user
        this.body = comment.body
        this.reactions = comment.reactions
    }

    static cast(comments: Readonly<IComment[]>): Comment[] {
        return comments.map((c: IComment): Comment => new Comment(c))
    }

    async isHeartBySelf(util: GithubKit): Promise<boolean> {
        if (this._existHeartReaction()) {
            return util.isHeartBySelf(this)
        }
        return Promise.resolve(false)
    }

    private _existHeartReaction(): boolean {
        return (this.reactions?.heart || 0) >= 1
    }
}
