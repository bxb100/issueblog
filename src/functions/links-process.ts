import * as core from '@actions/core'
import {Comment} from '../common/clazz/comment'
import {Config} from '../util/config'
import {Constant} from '../common/clazz/constant'
import {IComment} from '../common/interface/comment'
import {BlogContext} from '../common/clazz/blog-context'

// -----------------------------------------------------------------------------
/**
 * name:xxxxxx
 * link:xxxxxx
 * desc:xxxxxx
 */
// -----------------------------------------------------------------------------
export const FRIENDS_TABLE_HEAD =
    '| Name | Link | Desc |\n| ---- | ---- | ---- |\n'

export const friendTableTitle = (config: Config): string =>
    `\n## ${config.links_title}\n`
const friendTableTemplate = (
    name: string,
    link: string,
    desc: string
): string => `| ${name} | ${link} | ${desc} |\n`
const friendRegex = (str: string): RegExpExecArray | [] =>
    /(\w+):(.+)/.exec(str) || []

function _makeFriendTableString(comment: IComment): string {
    const dict: {[k: string]: string} = {}
    // eslint-disable-next-line github/array-foreach
    comment.body
        ?.split('\n')
        .filter(line => line.trim() !== '')
        .map(line => friendRegex(line))
        .filter(s => s !== null && s.length > 2)
        .forEach(s => (dict[s[1]] = s[2].trim()))

    if (dict) {
        core.debug(`_makeFriendTableString:\n\n${JSON.stringify(dict)}\n\n`)
        return friendTableTemplate(dict['name'], dict['link'], dict['desc'])
    }
    return ''
}

export async function add_md_friends(context: BlogContext): Promise<void> {
    const friendIssues = context.getIssues(Constant.FIXED_LINKS)

    const all: Promise<string[]>[] = []
    for (const issue of friendIssues) {
        all.push(
            context.kit
                .getIssueComments(issue)
                .then(async (comments: Comment[]) => {
                    const approved: IComment[] = []
                    for (const comment of comments) {
                        if (await comment.isHeartBySelf(context.kit)) {
                            approved.push(comment)
                        }
                    }
                    return approved
                })
                .then(approved => approved.map(_makeFriendTableString))
        )
    }
    const stringArray = await Promise.all(all).then(arr => arr.flat())
    if (stringArray.length <= 0) {
        return core.info("No friend's now.")
    }

    let friendSection: string = friendTableTitle(context.config)
    friendSection += FRIENDS_TABLE_HEAD
    friendSection += stringArray.join('')
    core.debug(`add_md_friends:\n\n${friendSection}\n\n`)

    context.sectionMap.set(Constant.FIXED_LINKS, friendSection)
}
