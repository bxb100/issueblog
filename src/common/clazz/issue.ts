import {GithubKit} from './github-kit'
import {IAssignee} from '../interface/assignee'
import {IIssue} from '../interface/issue'
import {ILabel} from '../interface/label'
import {IMilestone} from '../interface/milestone'
import {IUser} from '../interface/user'
import {IsoDateString} from '../types/iso-date-string'

export class Issue implements IIssue {
    readonly assignees: IAssignee[] | null | undefined
    readonly comments: number
    readonly created_at: IsoDateString
    readonly created_at_sub: string
    readonly labels: (string | ILabel)[]
    readonly locked: boolean
    readonly milestone: IMilestone | null | undefined
    readonly number: number
    readonly pull_request: Object | null | undefined
    readonly state: string
    readonly title: string
    readonly updated_at: IsoDateString
    readonly user: IUser | null
    readonly html_url: string
    readonly body: string | null | undefined

    constructor(data: IIssue) {
        this.assignees = data.assignees
        this.comments = data.comments
        this.created_at = data.created_at
        this.labels = data.labels
        this.locked = data.locked
        this.milestone = data.milestone
        this.number = data.number
        this.pull_request = data.pull_request
        this.state = data.state
        this.title = data.title
        this.updated_at = data.updated_at
        this.user = data.user
        this.html_url = data.html_url
        this.body = data.body

        // easy way get yyyy-MM-dd
        this.created_at_sub = this.created_at.substring(0, 10)
    }

    static cast(data: IIssue[]): Issue[] {
        return data.map(d => new Issue(d))
    }

    static getLabelValue(label: string | ILabel): string | undefined {
        if (typeof label === 'object') {
            return label.name
        }
        return label
    }

    getLabelName(kit: GithubKit): string[] {
        return this.labels
            .map(l => Issue.getLabelValue(l) || kit.config.unlabeled_title)
            .filter(Boolean)
    }

    containLabel(label: string): boolean {
        label = label.toLowerCase().trim()
        return this.labels.some(
            l => Issue.getLabelValue(l)?.toLowerCase() === label
        )
    }

    bodyToLines(): string[] {
        return this.body?.split('\n').map(line => line.trim()) || []
    }

    mdIssueInfo(): string {
        return `- [${this.title}](${this.html_url})---${this.created_at_sub}\n`
    }
}
