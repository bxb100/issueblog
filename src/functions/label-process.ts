import * as core from '@actions/core'
import {Constant} from '../common/clazz/constant'
import {Issue} from '../common/clazz/issue'
import {wrapDetails} from '../util/util'
import {BlogContext} from '../common/clazz/blog-context'

export async function add_md_label(context: BlogContext): Promise<void> {
    const issues = context.essayIssues

    const bucket: {[k: string]: Issue[]} = {}

    if (context.config.unlabeled_title) {
        bucket[context.config.unlabeled_title] = []
    }

    for (const issue of issues) {
        if (issue.labels.length === 0 && context.config.unlabeled_title) {
            bucket[context.config.unlabeled_title].push(issue)
            continue
        }
        for (const label of issue.labels) {
            const labelValue = Issue.getLabelValue(label)
            // skip save `top` label to bucket
            if (
                labelValue &&
                labelValue.toLowerCase() !== Constant.FIXED_TOP.toLowerCase()
            ) {
                if (!bucket[labelValue]) {
                    bucket[labelValue] = []
                }
                bucket[labelValue].push(issue)
            }
        }
    }
    const anchorNumber: number = parseInt(context.config.anchor_number)

    let labelSection = ''
    for (const key of Object.keys(bucket).sort((a, b) => a.localeCompare(b))) {
        labelSection += `\n## ${key}\n`
        const issueList = bucket[key]

        labelSection += wrapDetails(
            issueList.slice(0, anchorNumber),
            issueList.slice(anchorNumber),
            i => i.mdIssueInfo()
        )
    }
    core.debug(`labelSection: ${labelSection}`)
    context.sectionMap.set(Constant.AGG_EACH_LABEL, labelSection)
}
