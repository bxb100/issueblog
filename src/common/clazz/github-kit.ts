import * as core from '@actions/core'
import {context, getOctokit} from '@actions/github'
import {Comment} from './comment'
import {Config} from '../../util/config'
import {Constant} from './constant'
import {GitHub} from '@actions/github/lib/utils'
import {IComment} from '../interface/comment'
import {IIssue} from '../interface/issue'
import {IRelease} from '../interface/release'
import {Issue} from './issue'
import {Reaction} from '../interface/reaction'
import {ReactionContent} from '../enum/reaction-content'
import {Release} from './release'
import {add_md_friends} from '../../functions/links-process'
import {add_md_label} from '../../functions/label-process'
import {add_md_recent} from '../../functions/recent-process'
import {add_md_todo} from '../../functions/todo-process'
import {add_md_top} from '../../functions/top-process'
import {backup} from '../../functions/backup'
import fs from 'fs'
import {rss} from '../../functions/rss'

export class GithubKit {
    readonly client: InstanceType<typeof GitHub>
    readonly owner: string
    readonly repo: string
    readonly config: Config
    readonly sectionMap = new Map<string, string>()

    constructor(config: Config) {
        this.config = config
        this.client = getOctokit(config.github_token)
        this.owner = context.repo.owner
        this.repo = context.repo.repo
    }

    async isHeartBySelf(comment: IComment): Promise<boolean> {
        const reactions = await this.getIssueCommentReactions(
            comment,
            ReactionContent.HEART
        )
        return !!reactions.find(r => r.user?.login === this.owner)
    }

    async getIssueCommentReactions(
        comment: IComment,
        content?: ReactionContent
    ): Promise<Reaction[]> {
        const reactions = await this.client.rest.reactions.listForIssueComment({
            owner: this.owner,
            repo: this.repo,
            comment_id: comment.id,
            content
        })
        core.debug(`reactions:\n\n${JSON.stringify(reactions)}\n\n`)
        return reactions.data
    }

    async getIssueComments(issue: IIssue): Promise<Comment[]> {
        const comments = await this.client.rest.issues.listComments({
            owner: this.owner,
            repo: this.repo,
            issue_number: issue.number,
            per_page: 100
        })
        core.debug(`comments:\n\n${JSON.stringify(comments)}\n\n`)
        return Comment.cast(comments.data)
    }

    async getIssues(page: number): Promise<Issue[]> {
        const issueResult = await this.client.rest.issues.listForRepo({
            owner: this.owner,
            repo: this.repo,
            state: 'open',
            creator: this.owner,
            per_page: 100,
            direction: 'desc',
            page
        })
        core.debug(`issueResult:\n\n${JSON.stringify(issueResult.data)}\n\n`)
        return Issue.cast(issueResult.data)
    }

    async getRelease(page: number): Promise<IRelease[]> {
        const releases = await this.client.rest.repos.listReleases({
            owner: this.owner,
            repo: this.repo,
            per_page: 100,
            page
        })
        core.debug(`releases:\n\n${JSON.stringify(releases)}\n\n`)
        return releases.data
    }

    async getAllReleases(): Promise<Release[]> {
        let page = 1
        let releases: IRelease[] = []
        // eslint-disable-next-line no-constant-condition
        while (true) {
            const result = await this.getRelease(page)
            if (result.length === 0) {
                break
            }
            releases = releases.concat(result)
            page++
        }
        return Release.cast(releases)
    }

    async renderMarkdown(markdown: string): Promise<string> {
        // noinspection ES6RedundantAwait,TypeScriptValidateJSTypes
        const rendered = await this.client.rest.markdown.render({
            text: markdown,
            mode: 'markdown'
        })
        core.debug(`rendered:\n\n${JSON.stringify(rendered)}\n\n`)
        return rendered.data
    }

    async getAllIssues(): Promise<Issue[]> {
        let page = 1
        let issues: Issue[] = []
        // eslint-disable-next-line no-constant-condition
        while (true) {
            const result = await this.getIssues(page)
            if (result.length === 0) {
                break
            }
            issues = issues.concat(result)
            page++
        }
        return issues
    }

    async process(): Promise<void[]> {
        // using subscriber-publisher rewrite this is better?
        const issues = await this.getAllIssues()

        const mainProcess: Promise<void> = Promise.all([
            add_md_friends(this, issues),
            add_md_top(this, issues),
            add_md_recent(this, issues),
            add_md_todo(this, issues),
            add_md_label(this, issues)
        ]).then(
            () => {
                const constant = new Constant(this.config.md_header)
                fs.writeFileSync(
                    'README.md',
                    constant.convertBlogContent(this.sectionMap)
                )
            },
            err => {
                throw err
            }
        )
        return Promise.all([
            mainProcess,
            rss(this, issues),
            backup(this, issues)
        ])
    }
}
