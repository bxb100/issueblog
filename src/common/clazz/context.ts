import {GithubKit} from './github-kit'
import {Constant} from './constant'
import {Issue} from './issue'
import {Config} from '../../util/config'

export class Context {
    readonly kit: GithubKit
    readonly issues: Issue[] = []
    readonly sectionMap = new Map<string, string>()
    readonly essayIssues: Issue[] = []
    readonly config: Config

    constructor(issues: Issue[], kit: GithubKit, config: Config) {
        this.kit = kit
        this.issues = issues
        this.essayIssues = Context.getEssayIssues(issues)
        this.config = config
    }

    getIssues(label: string): Issue[] {
        if (label) {
            return this.issues.filter(issue => issue.hasLabel(label))
        }
        return this.issues
    }

    private static getEssayIssues(issues: Issue[]): Issue[] {
        return issues.filter(
            issue =>
                !(
                    issue.hasLabel(Constant.FIXED_LINKS) ||
                    issue.hasLabel(Constant.FIXED_TODO)
                )
        )
    }
}
