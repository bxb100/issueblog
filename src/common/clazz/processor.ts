import {Context} from './context'
import {Constant} from './constant'
import fs from 'fs'
import {add_md_friends} from '../../functions/links-process'
import {add_md_top} from '../../functions/top-process'
import {add_md_recent} from '../../functions/recent-process'
import {add_md_todo} from '../../functions/todo-process'
import {add_md_label} from '../../functions/label-process'
import {rss} from '../../functions/rss'
import {files} from '../../functions/files'
import {GithubKit} from './github-kit'
import {Config} from '../../util/config'

export class Processor {
    readonly config: Config

    constructor(config: Config) {
        this.config = config
    }

    writeReadMe(context: Context): Processor {
        const constant = new Constant(this.config.md_header)
        fs.writeFileSync(
            'README.md',
            constant.convertBlogContent(context.sectionMap)
        )
        return this
    }

    async process(): Promise<Processor> {
        const context = await this.init()

        return Promise.all([
            add_md_friends(context),
            add_md_top(context),
            add_md_recent(context),
            add_md_todo(context),
            add_md_label(context)
        ]).then(
            () => this.writeReadMe(context),
            err => {
                throw err
            }
        )
    }

    async rss(): Promise<Processor> {
        await rss(await this.init())
        return this
    }

    async files(): Promise<Processor> {
        await files(await this.init())
        return this
    }

    deleteAllFiles(): Processor {
        fs.rmdirSync(this.config.save_md_path)
        return this
    }

    private context: Context | null = null

    private async init(): Promise<Context> {
        if (this.context == null) {
            const kit = new GithubKit(this.config.github_token)
            this.context = new Context(
                await kit.getAllIssues(),
                kit,
                this.config
            )
        }
        return this.context
    }
}
