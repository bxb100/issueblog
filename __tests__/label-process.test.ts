import {FRIENDS_TABLE_HEAD, friendTableTitle} from '../src/functions/links-process'
// @ts-ignore
import {utils} from './base.test'
import {Constant} from '../src/common/clazz/constant'
import {add_md_label} from '../src/functions/label-process'

test('test generate string', async () => {

    await add_md_label(utils, await utils.getIssues(1))
    console.log(utils.sectionMap.get(Constant.EACH_LABEL))

    expect(utils.sectionMap.get(Constant.EACH_LABEL)).not.toContain('links')
    // worst case
}, 10000)
