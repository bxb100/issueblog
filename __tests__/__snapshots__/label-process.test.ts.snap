// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test generate string 1`] = `
"
## 2021
<details><summary>MORE</summary>

- [测试隐藏 label 6](https://github.com/bxb100/issueblog-test/issues/10)---2021-12-04
- [测试隐藏 label 5](https://github.com/bxb100/issueblog-test/issues/9)---2021-12-04
- [测试隐藏 label 4](https://github.com/bxb100/issueblog-test/issues/8)---2021-12-04
- [测试隐藏 label 3](https://github.com/bxb100/issueblog-test/issues/7)---2021-12-04
- [测试隐藏 label 2](https://github.com/bxb100/issueblog-test/issues/6)---2021-12-04
- [测试隐藏 label 1](https://github.com/bxb100/issueblog-test/issues/5)---2021-12-04
</details>


## two
<details><summary>MORE</summary>

- [测试隐藏 label 1](https://github.com/bxb100/issueblog-test/issues/5)---2021-12-04
</details>


## 测试
<details><summary>MORE</summary>

- [测试一个文章](https://github.com/bxb100/issueblog-test/issues/21)---2021-12-05
</details>

"
`;
